servant-server/src/Servant/Server/Internal/ErrorFormatter.hs

```diff
@@ -0,0 +1,63 @@
+{-# LANGUAGE DataKinds #-}
+
+-- | TODO: more documentation for this module
+module Servant.Server.Internal.ErrorFormatter
+  where
+
+import           Data.String.Conversions
+                 (cs)
+import           Data.Typeable
+
+import           Servant.Server.Internal.Context
+import           Servant.Server.Internal.ServerError
+
+-- | 'Context' that contains default formatters for all error types.
+--
+-- Default formatters will just return HTTP 400 status code with error
+-- message as response body.
+type DefaultErrorFormatters = '[BodyParseErrorFormatter, URLParseErrorFormatter, HeaderParseErrorFormatter]
+
+defaultErrorFormatters :: Context DefaultErrorFormatters
+defaultErrorFormatters =
+  defaulyBodyParseErrorFormatter
+  :. defaultURLParseErrorFormatter
+  :. defaultHeaderParseErrorFormatter
+  :. EmptyContext
+
+-- | A custom formatter for errors produced by parsing combinators like
+-- 'Servant.API.ReqBody' or 'Servant.API.Capture'.
+--
+-- A 'TypeRep' argument described the concrete combinator that raised
+-- the error, allowing formatter to customize the message for different
+-- combinators.
+type ErrorFormatter = TypeRep -> String -> ServerError
```

Comment:
I think since servant supports returning response in multiple different formats (e.g. `Post [JSON, PlainText] Response`), being able to specify only one fixed error formatter might be a limitation. Maybe also make this type take the value of `Accept` header?

Reply:
Good point! That was brought up in #685, but I deliberately ignored it for now. I think that for most applications it would be enough to force one type in the formatters.

There is no sane way to get verb's content-types while parsing URL parts and the body, as far as I can tell. For example:

```
type API = ReqBody '[JSON] Foo :> ("json" :> Get '[JSON] Bar :<|> "text" :> Get '[PlainText] Bar)
```

How would you know what content-type to return while parsing the `ReqBody`?

On the other hand, it's sure possible to add current `Accept` header value to formatter's arguments, should not be too hard to do.

Let's hear what the maintainers think.

Reply:
If we were to pass `Accept` to error formatter, one could then write formatter like this:

```haskell
errorFormatter acc rep err = handleAcceptH (Proxy @'[JSON, PlainText]) acc $ object ["combinator" .= show rep, "error" .= err]
```

Modulo the need to wrap that into a `ServerError` of course.
Would that make sense, @fisx?

Reply:
> How would you know what content-type to return while parsing the ReqBody?

By inspecting the path in the request?

Passing `Accept` would work, yes.  Seems like a reasonable solution to me.

Reply:
Okay, I'll do it.